[HOP] Mutation and alias rework

bohnstingl opened the pull request:
This PR reworks the way the input mutations and various aliases are checked

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @ydwu4

bohnstingl committed:
WIP: rework

bohnstingl committed:
Lintrunner

bohnstingl commented:
@pytorchbot label "topic: not user facing"

bohnstingl committed:
Fixed import issue with FlexAttention

bohnstingl committed:
Code cleanup

bohnstingl committed:
Fixed missing import issues

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Integrated lifted arguments and aot_eager backend for associative_scan

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -5938,7 +5945,8 @@ def f(xs, y):
         with self.assertRaisesRegex(
-            UnsupportedAliasMutationException, "torch.map is mutating the input!"
+            UnsupportedAliasMutationException,
+            "A branch or combine_fn might be modifying the input!.*",
```
The error message for hops can be different. We need to make sure the error messsages makes sense.

bohnstingl replied:
I reworked the error messages

ydwu4 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -191,6 +210,40 @@ def _maybe_fake_tracing(fn, inputs: List[Any], pre_dispatch):
+
+
+def check_input_mutation_and_alias(combine_fn, sample_inputs, pre_dispatch=False):
+    _, has_branch_alias = check_input_mutation(
```
why not call has_potential_input_mutation_or_alias directly?  "check_input_mutation" no need to return has_branch_alias anymore, which is pretty weird.

bohnstingl replied:
I reworked that check and isolated the aliases and mutations such that it should be easier in a follow-up PR to determine which node is triggering the checks

bohnstingl committed:
WIP commit

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
WIP

bohnstingl committed:
Moved alias and mutation checks to dynamo for associative_scan and scan

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Integrated mutation check to cond

bohnstingl committed:
Reworked alias and mutation checks for scan, associative_scan, while_loop, cond and map

bohnstingl commented on torch/_dynamo/graph_deduplication.py:
```diff
@@ -120,6 +120,7 @@ def _replace_region_with_subgraph(
+    # TODO: We don't care here about any output-output aliasing?
     if has_potential_input_alias_or_mutation(sub_gm, fake_inputs):
         log.debug(
             "NYI: Failed to substitute region %s due to input alias or mutation",
```
I revised

bohnstingl commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -2957,6 +2957,7 @@ def call_function(
+        # TODO: We don't care here about any output-output aliasing?
         if has_potential_input_alias_or_mutation(body_gmod, fake_inputs):
             raise RuntimeError(
                 f"{self.value._name} where the inputs are mutated or the "
```
I revised

bohnstingl commented:
@ydwu4 I reworked the mutation and alias checks. I moved the checks into dynamo for scan, associative_scan, while_loop and cond. For map I also included the new check, but since it does not yet use backend='eager', I did not move the check to dynamo. Pleas let me know what you think

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -127,6 +126,55 @@ def _unwrap_var(var):
 
     return check_meta_consistency(unwrapped1, unwrapped2, lhs_name, rhs_name)
 
+def flatten_nested_lists(li):
```
why not pytree.tree_flatten?

bohnstingl replied:
Sure, I changed that.

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -127,6 +126,55 @@ def _unwrap_var(var):
+                            val = torch._C._functorch.get_unwrapped(val)
+                        inputs_fake.append(val)
+                    else:
+                        # This is the standard case of a TensorVariable
```
can assert it's FakeTensor?

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -127,6 +126,55 @@ def _unwrap_var(var):
+                    # This case is for SymInts and other non-Tensor elements
+                    inputs_fake.append(val)
+            else:
+                # This case is for ints
```
can assert they're ints?

bohnstingl replied:
Done.

I was wondering though in general whether we could somehow improve this? Are you fine with we currently collect the fake inputs?

ydwu4 commented on torch/_higher_order_ops/flex_attention.py:
```diff
@@ -7,13 +7,12 @@
-    _has_potential_branch_input_mutation,
     _maybe_reenter_make_fx,
     autograd_not_implemented,
+    # check_input_mutation,
```
remove?

bohnstingl replied:
Sure, I updated the PR and now reworked pretty much all HOPs

ydwu4 commented on torch/_higher_order_ops/hints_wrap.py:
```diff
@@ -3,12 +3,10 @@
-    _has_potential_branch_input_alias,
-    _has_potential_branch_input_mutation,
     autograd_not_implemented,
+    # check_input_mutation_and_alias,
```
remove?

bohnstingl replied:
See above.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -442,27 +439,6 @@ def scan_functionalize(ctx, combine_fn, init, xs, additional_inputs):
     unwrapped_additional_inputs = ctx.unwrap_tensors(additional_inputs)
     with ctx.redispatch_to_next():
         functional_combine_fn = ctx.functionalize(combine_fn)
```
we might still want to keep the check in functionalization key. In case someone is using the hop directly, which could bypass the dynamo checks.

bohnstingl replied:
Right, as discussed offline, I reintroduced the checks into the `functionalization` key as well. 
If the HOP uses the backend='eager', we now we have it in `functionalization` key and in `dynamo`.

ydwu4 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -201,43 +173,26 @@ def has_potential_input_alias_or_mutation(gm, inputs, pre_dispatch=False):
+                          inp_out_alias_map,
+                          out_out_alias_map)
+
+def has_potential_input_mutation_or_alias(gm, inputs, pre_dispatch=False):
```
nit: Probably we should still name this as has_potential_input_alias_or_mutation?  The name change seems unnecessary

bohnstingl replied:
I agree. I corrected it.

ydwu4 reviewed (commented):
Looks good overall. Left a few minor comments.

bohnstingl committed:
WIP

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Integrated code reviews

ydwu4 commented on torch/_higher_order_ops/while_loop.py:
```diff
@@ -467,8 +467,8 @@ def while_loop_func(ctx, cond_fn, body_fn, carried_inputs, additional_inputs):
         for fn, fn_name in [
-            (cond_fn, "cond_fn"),
-            (body_fn, "body_fn"),
+            (functional_cond_fn, "cond_fn"),
```
why we need to change this?

bohnstingl replied:
I guess we don't have to use the functionalized variable. I reverted it and also for all the other HOPs we use the non functionalized variable for the mutation and alias checking.

bohnstingl committed:
Fixed lint issues
Corrected naming and return arguments of functions
Integrated code review comment

bohnstingl committed:
Fixed lint issues

bohnstingl committed:
WIP

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Deferred flex_attention mutation checking

bohnstingl committed:
Fixed import issue with flex_attention

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Only use input mutation for flex_attention

bohnstingl committed:
Fixed variable name error

bohnstingl committed:
Fixed imports

bohnstingl committed:
Fixed issues with mutation and alias check, i.e., pytree output needed to be flattened

bohnstingl committed:
Fixed alias issue with cond

bohnstingl committed:
Fixed issues with CI tests for input mutation/alias. Reverted checks for auto_functionalization

bohnstingl commented:
Currently we have only limited alias and mutation checks for `flex_attention`, i.e., we only check for input mutation and we also don't check either input alias or input mutation for `auto_functionalize`.

bohnstingl committed:
Unified analysis of mutation and alias

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Fixed testcases

bohnstingl committed:
Lintrunner fixes

bohnstingl committed:
Fixed some more testcases and cleaned resolved issue references

bohnstingl committed:
Removed unnecessary code

bohnstingl committed:
Update to executorch_call_delegate

bohnstingl committed:
Fixed issue with lifted arguments in mutation checks for scan and associative_scan

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl commented:
There is a potential overlap with https://github.com/pytorch/pytorch/pull/148953 and it should be discussed what changes from here are relevant and should be kept and what is redundant.

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Updates to HOPs and to alias checking speculate_subgraph

bohnstingl committed:
Fixed issues with new alias rework and fake inputs

bohnstingl commented:
@ydwu4, @anijain2305: I updated this PR given the this [PR](https://github.com/pytorch/pytorch/pull/148953) and the discussion [here](https://github.com/pytorch/pytorch/issues/149989). What do you think about it?

bohnstingl committed:
Fixing lint issues and CI tests

bohnstingl committed:
Fixed further CI testcases and lint issues

bohnstingl committed:
Fixed cond testcase

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Reverted executorch HOP

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Fixed CI tests

bohnstingl committed:
Fixed CI tests

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Removed unnecessary code

bohnstingl committed:
Fixed merge issues

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Fixed CI tests

bohnstingl committed:
Update invoke_subgraph graphs

bohnstingl committed:
Reverted invoke_subgraph and FunctionalizeCtxWrapper

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -888,8 +888,12 @@ def forward(self, arg0_1, arg1_1, arg2_1, arg3_1, arg4_1, arg5_1, arg6_1):
 
     def test_cond_autograd_pytree_input(self):
+        # TODO: This is an unexpected behavior for cond
+        # Without this additional multiplication,
+        # the output of the backward graph would alias the
+        # inputs, as the gradients are just 1s and thus get optimized
         def true_fn(x):
-            return x["t"][0] + x["t"][1]["b"] * x["t"][2][0]
+            return (x["t"][0] * 2.0) + x["t"][1]["b"] * x["t"][2][0]
```
This may be a bit unexpected for to the user. Currently we don't allow aliases, including input-output aliases. This is problematic, because the gradients could be just 1s, in which case the gradients (arguments) from the upstream are just passed along, which then triggers the alias checks.

A naive solution would be to disable the input-output alias check, but I am not sure whether this causes problems? 
Is there another solution to this?

ydwu4 replied:
A better way is to properly support it through auto_functionalized. This is still WIP though.

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Fixed issue with return order of alias and mutation function

bohnstingl committed:
Fixed cond testcases

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Cleaned up code

ydwu4 commented on torch/_dynamo/output_graph.py:
```diff
@@ -2831,11 +2831,13 @@ def has_input_mutation(self):
         for node in self.graph.nodes:
             if node.op == "placeholder":
-                example_value = node.meta["example_value"]
+                example_value = _collect_fake_inputs([node])[0]
```
what happens here? I don't expect it will error, will it?

bohnstingl replied:
No, it does error out. The issue is that in some testcases. For example in, the issue is that the example value is a `BatchedTensorImpl`, for which `self.untyped_storage()` doesn't exist.

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1336,6 +1340,8 @@ def create_unbacked_sym_node_var(tx) -> SymNodeVariable:
             source_target=self.value,
             set_subgraph_inputs="flatten_manual",
             should_flatten_outputs=True,
+            supports_input_mutation=False,
```
Can put these two as class field? like BaseHOPVariable

bohnstingl replied:
Done.

ydwu4 commented on torch/_higher_order_ops/utils.py:
```diff
@@ -699,17 +741,19 @@ def validate_subgraph_args_types(lifted_args: Union[tuple[Any, ...], list[Any]])
 
+# TODO: Return a more detailed information as to which node
+# causes a mutation or an alias. This may requires a per operator tensor version checking
 def check_input_alias_and_mutation(
```
why we need to move mutated_inputs to the end?

bohnstingl replied:
Well, I just moved it in order to be consistent with the function name. It has alias and mutation and that's why I moved it towards the end. WDYT?

ydwu4 reviewed (commented):
Tests are failing. Added a few comments

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Fixed some CI issues and integrated review comments

ydwu4 reviewed (approved):
The change looks good. Not sure why the test starts to fail.

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into mutation_alias_rework

bohnstingl committed:
Fixes to CI tests because of alias_mutation order

bohnstingl commented:
I think I found the issue. There was one occasion in the inductor where I missed the rearrangement of the return variable from alias_mutation. Fingers crossed for this time.

ydwu4 commented:
@pytorchbot merge
